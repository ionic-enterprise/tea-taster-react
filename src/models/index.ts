export * from './Tea';
export * from './User';
export * from './Session';
export * from './TastingNote';
export * from './UnlockMode';
