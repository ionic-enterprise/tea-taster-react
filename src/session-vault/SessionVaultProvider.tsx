import { ReactNode, createContext, useContext, useEffect, useMemo, useState } from 'react';
import { useIonModal } from '@ionic/react';
import { PinDialog } from './PinDialog';
import { clearSessionCache, vault } from '../api/session-vault-api';
import { useHistory } from 'react-router';

type Props = { children?: ReactNode };
type CustomPasscodeCallback = (opts: { data: any; role?: string }) => void;
let handlePasscodeRequest: CustomPasscodeCallback = () => {};

const SessionVaultContext = createContext<any>({});
const SessionVaultProvider = ({ children }: Props) => {
  const history = useHistory();
  const [showModal, setShowModal] = useState<boolean>(false);
  const [isSetPasscodeMode, setIsSetPasscodeMode] = useState<boolean>(false);
  const [present, dismiss] = useIonModal(PinDialog, {
    setPasscodeMode: isSetPasscodeMode,
    onDismiss: (opts: { data: any; role?: string }) => handlePasscodeRequest(opts),
  });

  useMemo(() => {
    vault.onLock(() => {
      clearSessionCache();
      history.replace('/unlock');
    });
    vault.onPasscodeRequested(async (isPasscodeSetRequest) => {
      return new Promise((resolve) => {
        handlePasscodeRequest = (opts) => {
          vault.setCustomPasscode(opts.role === 'cancel' ? '' : opts.data);
          setIsSetPasscodeMode(false);
          setShowModal(false);
          resolve();
        };
        setIsSetPasscodeMode(isPasscodeSetRequest);
        setShowModal(true);
      });
    });
  }, []);

  useEffect(() => {
    showModal ? present() : dismiss();
  }, [showModal]);

  return <SessionVaultContext.Provider value={{}}>{children}</SessionVaultContext.Provider>;
};
export const useSessionVault = () => {
  const context = useContext(SessionVaultContext);
  if (context === undefined) throw new Error('useSessionVault must be used within SessionVaultContext');
  return context;
};
export default SessionVaultProvider;
