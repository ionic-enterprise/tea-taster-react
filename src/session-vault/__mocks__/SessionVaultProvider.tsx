import { vi } from 'vitest';

const canUnlock = vi.fn().mockResolvedValue(undefined);
const getUnlockMode = vi.fn().mockResolvedValue(undefined);
const setUnlockMode = vi.fn().mockResolvedValue(undefined);

export const useSessionVault = () => ({ canUnlock, getUnlockMode, setUnlockMode });

const SessionVaultProvider = ({ children }: any) => <div>{children}</div>;
export default SessionVaultProvider;
