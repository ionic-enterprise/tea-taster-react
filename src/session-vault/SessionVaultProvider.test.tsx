import { vi, Mock } from 'vitest';
import { Session, UnlockMode } from '../models';
import SessionVaultProvider, { useSessionVault } from './SessionVaultProvider';

vi.mock('@capacitor/preferences');
vi.mock('../api/vault-factory-api');

const MockChildComponent = () => {
  return <div data-testid="session-vault"></div>;
};

const mockComponent = (
  <SessionVaultProvider>
    <MockChildComponent />
  </SessionVaultProvider>
);

describe('SessionVaultProvider', () => {
  const wrapper = ({ children }: any) => <SessionVaultProvider>{children}</SessionVaultProvider>;

  let mockVault: any;
  const testSession: Session = {
    user: { id: 1, firstName: 'John', lastName: 'Doe', email: 'john.doe@test.com' },
    token: '123456789',
  };

  beforeEach(() => vi.clearAllMocks());
});
