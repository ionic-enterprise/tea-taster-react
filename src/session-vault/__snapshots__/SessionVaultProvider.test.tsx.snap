// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SessionVaultProvider > renders consistently 1`] = `<DocumentFragment />`;
